File: src/main/java/net/minestom/server/utils/binary/PooledBuffers.java
package net.minestom.server.utils.binary;

import net.minestom.server.network.socket.Server;
import net.minestom.server.utils.cache.LocalCache;
import org.jctools.queues.MessagePassingQueue;
import org.jctools.queues.MpmcUnboundedXaddArrayQueue;
import org.jetbrains.annotations.ApiStatus;

import java.lang.ref.Cleaner;
import java.lang.ref.SoftReference;
import java.nio.ByteBuffer;
import java.util.Collection;
import java.util.concurrent.atomic.AtomicReference;

@ApiStatus.Internal
@ApiStatus.Experimental
public final class PooledBuffers {
    private final static MessagePassingQueue<SoftReference<BinaryBuffer>> POOLED_BUFFERS = new MpmcUnboundedXaddArrayQueue<>(1024);
    private final static int BUFFER_SIZE = Integer.getInteger("minestom.pooled-buffer-size", 262_143);
    private final static Cleaner CLEANER = Cleaner.create();

    private static final LocalCache<ByteBuffer> PACKET_BUFFER = LocalCache.ofBuffer(Server.MAX_PACKET_SIZE);
    private static final LocalCache<ByteBuffer> LOCAL_BUFFER = LocalCache.ofBuffer(Server.MAX_PACKET_SIZE);

    /**
     * Thread local buffer containing raw packet stream.
     */
    public static ByteBuffer packetBuffer() {
        return PACKET_BUFFER.get().clear();
    }

    /**
     * Thread local buffer targeted at very small scope operations (encryption, compression, ...).
     */
    public static ByteBuffer tempBuffer() {
        return LOCAL_BUFFER.get().clear();
    }

    public static BinaryBuffer get() {
        BinaryBuffer buffer;
        SoftReference<BinaryBuffer> ref;
        while ((ref = POOLED_BUFFERS.relaxedPoll()) != null) {
            if ((buffer = ref.get()) != null) return buffer;
        }
        return BinaryBuffer.ofSize(BUFFER_SIZE);
    }

    public static void add(BinaryBuffer buffer) {
        POOLED_BUFFERS.relaxedOffer(new SoftReference<>(buffer.clear()));
    }

    public static void clear() {
        POOLED_BUFFERS.clear();
    }

    public static int count() {
        return POOLED_BUFFERS.size();
    }

    public static int bufferSize() {
        return BUFFER_SIZE;
    }

    public static void registerBuffer(Object ref, AtomicReference<BinaryBuffer> buffer) {
        CLEANER.register(ref, new BufferRefCleaner(buffer));
    }

    public static void registerBuffer(Object ref, BinaryBuffer buffer) {
        CLEANER.register(ref, new BufferCleaner(buffer));
    }

    public static void registerBuffers(Object ref, Collection<BinaryBuffer> buffers) {
        CLEANER.register(ref, new BuffersCleaner(buffers));
    }

    private record BufferRefCleaner(AtomicReference<BinaryBuffer> bufferRef) implements Runnable {
        @Override
        public void run() {
            add(bufferRef.get());
        }
    }

    private record BufferCleaner(BinaryBuffer buffer) implements Runnable {
        @Override
        public void run() {
            add(buffer);
        }
    }

    private record BuffersCleaner(Collection<BinaryBuffer> buffers) implements Runnable {
        @Override
        public void run() {
            if (buffers.isEmpty()) return;
            for (BinaryBuffer buffer : buffers) {
                add(buffer);
            }
        }
    }
}

File: src/main/java/net/minestom/server/utils/binary/PooledBuffers.kt
package net.minestom.server.utils.binary

import org.jglrxavpok.hephaistos.nbt.NBTReader.read
import org.jglrxavpok.hephaistos.nbt.NBTWriter.writeNamed
import net.minestom.server.utils.binary.BinaryReader
import net.minestom.server.utils.binary.BinaryWriter
import org.jglrxavpok.hephaistos.nbt.NBTReader
import org.jglrxavpok.hephaistos.nbt.NBTWriter
import net.minestom.server.utils.binary.BinaryBuffer
import java.lang.RuntimeException
import java.io.IOException
import java.nio.channels.WritableByteChannel
import java.nio.channels.ReadableByteChannel
import java.nio.BufferUnderflowException
import net.minestom.server.utils.SerializerUtils
import net.minestom.server.item.ItemStack
import net.minestom.server.utils.NBTUtils
import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer
import org.jglrxavpok.hephaistos.nbt.NBT
import org.jglrxavpok.hephaistos.nbt.CompressedProcesser
import net.minestom.server.MinecraftServer
import net.minestom.server.network.socket.Server
import org.jglrxavpok.hephaistos.nbt.NBTException
import java.lang.Runnable
import java.nio.charset.Charset
import net.minestom.server.utils.binary.Writeable
import java.util.function.BiConsumer
import java.util.function.ToIntFunction
import org.jctools.queues.MessagePassingQueue
import org.jctools.queues.MpmcUnboundedXaddArrayQueue
import net.minestom.server.utils.binary.PooledBuffers
import java.util.concurrent.atomic.AtomicReference
import net.minestom.server.utils.binary.PooledBuffers.BufferRefCleaner
import net.minestom.server.utils.binary.PooledBuffers.BufferCleaner
import net.minestom.server.utils.binary.PooledBuffers.BuffersCleaner
import net.minestom.server.utils.cache.LocalCache
import org.jetbrains.annotations.ApiStatus
import java.lang.ref.Cleaner
import java.lang.ref.SoftReference
import java.nio.ByteBuffer

@ApiStatus.Internal
@ApiStatus.Experimental
object PooledBuffers {
    private val POOLED_BUFFERS: MessagePassingQueue<SoftReference<BinaryBuffer?>> = MpmcUnboundedXaddArrayQueue(1024)
    private val BUFFER_SIZE = Integer.getInteger("minestom.pooled-buffer-size", 262143)
    private val CLEANER = Cleaner.create()
    private val PACKET_BUFFER = LocalCache.ofBuffer(Server.MAX_PACKET_SIZE)
    private val LOCAL_BUFFER = LocalCache.ofBuffer(Server.MAX_PACKET_SIZE)

    /**
     * Thread local buffer containing raw packet stream.
     */
    fun packetBuffer(): ByteBuffer {
        return PACKET_BUFFER.get().clear()
    }

    /**
     * Thread local buffer targeted at very small scope operations (encryption, compression, ...).
     */
    fun tempBuffer(): ByteBuffer {
        return LOCAL_BUFFER.get().clear()
    }

    fun get(): BinaryBuffer {
        var buffer: BinaryBuffer
        var ref: SoftReference<BinaryBuffer>
        while (POOLED_BUFFERS.relaxedPoll().also { ref = it } != null) {
            if (ref.get().also { buffer = it!! } != null) return buffer
        }
        return BinaryBuffer.Companion.ofSize(BUFFER_SIZE)
    }

    fun add(buffer: BinaryBuffer) {
        POOLED_BUFFERS.relaxedOffer(SoftReference(buffer.clear()))
    }

    fun clear() {
        POOLED_BUFFERS.clear()
    }

    fun count(): Int {
        return POOLED_BUFFERS.size()
    }

    fun bufferSize(): Int {
        return BUFFER_SIZE
    }

    fun registerBuffer(ref: Any?, buffer: AtomicReference<BinaryBuffer?>?) {
        CLEANER.register(ref, BufferRefCleaner(buffer))
    }

    fun registerBuffer(ref: Any?, buffer: BinaryBuffer?) {
        CLEANER.register(ref, BufferCleaner(buffer))
    }

    fun registerBuffers(ref: Any?, buffers: Collection<BinaryBuffer?>?) {
        CLEANER.register(ref, BuffersCleaner(buffers))
    }

    private inner class BufferRefCleaner : Runnable {
        override fun run() {
            add(bufferRef.get())
        }
    }

    private inner class BufferCleaner : Runnable {
        override fun run() {
            add(buffer)
        }
    }

    private inner class BuffersCleaner : Runnable {
        override fun run() {
            if (buffers.isEmpty()) return
            for (buffer in buffers) {
                add(buffer)
            }
        }
    }
}